Selecting an Association in the tree generates an error

### Describe the bug
When a programmatically created association is selected in the tree, neither the client or supplier values are displayed in the Property Editor

### To Reproduce
Steps to reproduce the behavior:
1. Create an association using the API. Set owning package, both memberEnds and ownedEnd 
2. Select the association in the user interface tree. An error is generated

### Expected behavior
The Property Editor should show the details of the association without showing an error.

### Screenshots
```
Gaphor version:         2.26.0
Operating System:       Windows (10)
Display:                GdkWin32Display
Python version:         3.12.4
GTK version:            4.14.4
Adwaita version:        1.5.2
GtkSourceView version:  5.12.1
Cairo version:          1.18.0
Pango version:          1.54.0
PyGObject version:      3.48.2
Pycairo version:        1.26.1


Errors:

Time since application startup: 0:23:27
|Traceback (most recent call last):
|  File "c:\Users\pbrow\Workspaces\GaphorMDImportPlugin\.venv\Lib\site-packages\gaphor\UML\classes\associationpropertypages.py", line 148, in _on_end_navigability_change
|    with Transaction(self.event_manager):
|  File "c:\Users\pbrow\Workspaces\GaphorMDImportPlugin\.venv\Lib\site-packages\gaphor\transaction.py", line 58, in __init__
|    self._handle(TransactionBegin(self.context))
|    |            |                -> <gaphor.transaction.Transaction object at 0x000002A8D3750B00>
|    |            -> <class 'gaphor.event.TransactionBegin'>
|    -> <gaphor.transaction.Transaction object at 0x000002A8D3750B00>
|  File "c:\Users\pbrow\Workspaces\GaphorMDImportPlugin\.venv\Lib\site-packages\gaphor\transaction.py", line 113, in _handle
|    self.event_manager.handle(event)
|    |                         -> <gaphor.event.TransactionBegin object at 0x000002A8D3750B30>
|    -> <gaphor.transaction.Transaction object at 0x000002A8D3750B00>
|  File "c:\Users\pbrow\Workspaces\GaphorMDImportPlugin\.venv\Lib\site-packages\gaphor\core\eventmanager.py", line 77, in handle
|    self._priority.handle(event)
|    |                     -> <gaphor.event.TransactionBegin object at 0x000002A8D3750B30>
|    -> <gaphor.core.eventmanager.EventManager object at 0x000002A8C9680380>
|  File "c:\Users\pbrow\Workspaces\GaphorMDImportPlugin\.venv\Lib\site-packages\generic\event.py", line 63, in handle
|    raise ExceptionGroup("Error while handling events", exceptions)
|                                                        -> [AssertionError()]
|ExceptionGroup: Error while handling events (1 sub-exception)
└─┬──────────────────────────────╌┄┈
  |Traceback (most recent call last):
  |  File "c:\Users\pbrow\Workspaces\GaphorMDImportPlugin\.venv\Lib\site-packages\generic\event.py", line 59, in handle
  |    handler(event)
  |    |       -> <gaphor.event.TransactionBegin object at 0x000002A8D3750B30>
  |    -> <bound method UndoManager.begin_transaction of <gaphor.services.undomanager.UndoManager object at 0x000002A8C9610F80>>
  |  File "c:\Users\pbrow\Workspaces\GaphorMDImportPlugin\.venv\Lib\site-packages\gaphor\services\undomanager.py", line 179, in begin_transaction
  |    assert not self._current_transaction
  |               -> <gaphor.services.undomanager.UndoManager object at 0x000002A8C9610F80>
  |AssertionError: assert not self._current_transaction
  └─────────────────────────────╌┄┈
```

### OS
- [ ] Linux (Please put in notes the specific distro)
- [ ] macOS
- [x] Windows

### Version
Version of Gaphor: 2.26.0

### Additional information
Add any other context about the problem here.

--- a/gaphor/UML/classes/associationpropertypages.py
+++ b/gaphor/UML/classes/associationpropertypages.py
@@ -1,3 +1,6 @@
+import contextlib
+import functools
+
 from gaphor import UML
 from gaphor.core.format import format, parse
 from gaphor.diagram.propertypages import (
@@ -14,6 +17,25 @@ from gaphor.UML.profiles.stereotypepropertypages import (
     stereotype_model,
     stereotype_set_model_with_interaction,
 )
+
+
+def blockable(func):
+    blocked = 0
+
+    def _blockable(*args, **kwargs):
+        if not blocked:
+            return func(*args, **kwargs)
+
+    @contextlib.contextmanager
+    def block():
+        nonlocal blocked
+        blocked += 1
+        yield
+        blocked -= 1
+
+    wrapped = functools.update_wrapper(_blockable, func)
+    wrapped.block = block  # type: ignore[attr-defined]
+    return wrapped
 
 
 @PropertyPages.register(UML.Association)
@@ -37,10 +59,12 @@ class AssociationPropertyPage(PropertyPageBase):
         self.update_end_name(builder, end_name, subject)
 
         navigation = builder.get_object(f"{end_name}-navigation")
-        navigation.set_selected(self.NAVIGABILITY.index(subject.navigability))
+        with self._on_end_navigability_change.block():
+            navigation.set_selected(self.NAVIGABILITY.index(subject.navigability))
 
         aggregation = builder.get_object(f"{end_name}-aggregation")
-        aggregation.set_selected(self.AGGREGATION.index(subject.aggregation))
+        with self._on_end_aggregation_change.block():
+            aggregation.set_selected(self.AGGREGATION.index(subject.aggregation))
 
         if stereotypes_model := stereotype_model(subject, self.event_manager):
             stereotype_list = builder.get_object(f"{end_name}-stereotype-list")
@@ -49,6 +73,7 @@ class AssociationPropertyPage(PropertyPageBase):
             stereotype_frame = builder.get_object(f"{end_name}-stereotype-frame")
             stereotype_frame.set_visible(False)
 
+    @blockable
     def update_end_name(self, builder, end_name, subject):
         name = builder.get_object(f"{end_name}-name")
         new_name = (
@@ -60,10 +85,8 @@ class AssociationPropertyPage(PropertyPageBase):
             )
             or ""
         )
-        if not (name.is_focus() or self.end_name_change_semaphore):
-            self.end_name_change_semaphore += 1
+        with self._on_end_name_change.block():
             name.set_text(new_name)
-            self.end_name_change_semaphore -= 1
         return name
 
     def construct(self):
@@ -136,13 +159,13 @@ class AssociationPropertyPage(PropertyPageBase):
             builder.get_object("association-editor"), self.watcher
         )
 
+    @blockable
     def _on_end_name_change(self, entry, subject):
-        if not self.end_name_change_semaphore:
-            self.end_name_change_semaphore += 1
+        with self.update_end_name.block():
             with Transaction(self.event_manager):
                 parse(subject, entry.get_text())
-            self.end_name_change_semaphore -= 1
 
+    @blockable
     def _on_end_navigability_change(self, dropdown, _pspec, subject):
         if subject and subject.opposite and subject.opposite.type:
             with Transaction(self.event_manager):
@@ -152,6 +175,7 @@ class AssociationPropertyPage(PropertyPageBase):
                     self.NAVIGABILITY[dropdown.get_selected()],
                 )
 
+    @blockable
     def _on_end_aggregation_change(self, dropdown, _pspec, subject: UML.Property):
         aggregation = self.AGGREGATION[dropdown.get_selected()]
         if aggregation != subject.aggregation:

